thecartercenter/nemo
# 12029 enketo editor!

## Bird's eye summary

resources:
- [enketo "getting started" guide](https://enketo.github.io/enketo-core/tutorial-00-getting-started.html)
- [enketo-transformer docs](https://github.com/enketo/enketo-transformer)

architecture/flow:
- nemo talks to `lib/enketo-transformer-service/index.js` (separate node app that runs enketo-transformer) to convert our form XML into a JSON object we're able to send to enketo
- nemo renders a `<div id="enketo">` in `enketo_form.html.erb`
- nemo injects 2 variables into JS via a `<script>` tag:
  - `window.ENKETO_MODEL_STR` is the form XML from enketo-transformer-service
  - `window.ENKETO_INSTANCE_STR` is the optional response XML (when editing)
- enketo injects itself into `document.querySelector('#enketo form')` via jQuery in `packs/enketo.js`
- enketo loads the two variables set above and renders the form
- user types in info and hits submit
- enketo submits XML to nemo's existing ODK `/submission` (`responses#create`) route
- on edit, nemo saves the edited XML to `modified_odk_xml` attachment so as to preserve the original `odk_xml` as required by some orgs

## Code review notes

main feedback requested from PR review:
- usability
- bugs
- architecture and general code review

things you can ignore:
- vendor SASS & JS files

nemo features to test:
- submit new response
- switch back and forth between web/enketo editor based on user preference
- edit response submitted via enketo
- edit response submitted via collect
- edit legacy response (response that was submitted earlier, then form was returned to draft and structurally modified; this should generally work without issue except in rare cases)
- error message when editing response submitted via web
- warning message when editing response (via enketo) most recently edited by web
- warning message when editing response (via web) most recently edited by enketo
- informative error popup when submitting fails (e.g. if your internet shuts off before you try to submit)
- unauthorized page, e.g. when trying to view someone else's response as an enumerator
- "powered by" footer whenever enketo is used
- weird symbols including " and ` in both questions and answers
- various question types (see below for full list)
- please also report any console errors you come across: enketo is written in jQuery so that's where errors will appear

## Release notes

enketo features to try out:
- skip logic ✨
- repeat groups
- internationalization (via NEMO interface)

question types that are supported already:
- Text
- Long Text
- Integer
- Barcode
- Counter
- Decimal
- Location (google maps does not yet have API hooked up)
- Select One (including cascading select)
- Select Multiple
- Date/Time (currently unstyled but functional)
- Date (currently unstyled but functional)
- Time (currently unstyled but functional)

media question types are not yet viewable/editable in enketo:
- Image
- Annotated Image
- Signature
- Sketch
- Audio
- Video

features coming later as we prioritize them:
- usability polish
- edit response submitted by web (requires nemo hierarchy => XML transformer)
- convert JS popup alerts (e.g. on error) into DOM elements (e.g. like rails "flash")
- support for media (e.g. file attachments in answers, image prompts in questions)
- even more detailed specs; may be changing based on usability feedback

features that already exist in nemo editor that we can eventually port to enketo editor as we prioritize them
- check out/in logic when multiple users are editing
- metadata section (that shows things like "is reviewed")
